workflow-runtime/src/main/java/com/squareup/workflow1/WorkflowInterceptor.kt

```diff
@@ -120,6 +124,146 @@ public interface WorkflowInterceptor {
     /** The parent [WorkflowSession] of this workflow, or null if this is the root workflow. */
     public val parent: WorkflowSession?
   }
+
+  /**
+   * Provides hooks for intercepting calls to a [BaseRenderContext], to be used from [onRender].
+   * See [interceptRenderContext] for more information.
+   */
+  public interface RenderContextInterceptor<P, S, O> {
+
+    /**
+     * Intercepts calls to [send][Sink.send] on the [BaseRenderContext.actionSink].
+     *
+     * This method will be called from inside the actual [Sink.send] stack frame, so any stack
+     * traces captured from it will include the code that is actually making the send call.
+     */
+    public fun onActionSent(
+      action: WorkflowAction<P, S, O>,
+      proceed: (WorkflowAction<P, S, O>) -> Unit
+    ) {
+      proceed(action)
+    }
+
+    /**
+     * Intercepts side effects that are ran via [BaseRenderContext.runningSideEffect].
+     *
+     * This method will only be called when a particular side effect is _started_ for a particular
+     * key, and will not be called on subsequent render passes that run the side effect with the
+     * same key.
+     *
+     * The [proceed] function will perform the actual suspending side effect, and only return when
+     * the side effect is complete – this may be far in the future. This means the interceptor can
+     * be notified when the side effect _ends_ by simply running code after [proceed] returns or
+     * throws.
+     *
+     * The interceptor may run [proceed] in a different [CoroutineContext], but the context _must_
+     * have the same [Job] or an exception will be thrown. This is to ensure that structured
+     * concurrency is not broken. The context can change things like dispatcher, name, etc.
+     */
+    public suspend fun onSideEffectRunning(
+      key: String,
+      proceed: suspend () -> Unit,
```

Comment:
These all being suspend functions makes a really simple API that is really powerful. I'm pretty excited about the idea of "suspending interceptors".